rustls/src/tls13/mod.rs

```diff
@@ -233,6 +261,41 @@ impl MessageDecrypter for Tls13MessageDecrypter {
     }
 }
 
+#[cfg(feature = "secret_extraction")]
+fn chacha20_poly1305_extract_keys(key: AeadKey, iv: Iv) -> ConnectionTrafficSecrets {
```

Comment:
I forget whether this is the same argument I made in my previous round of feedback, but it still feels like this means `AeadAlgorithm` should be more like a trait with an `extract_keys()` method? Why doesn't that work here?

Reply:
eventually we need 3 values that `impl Tls13AeadAlgorithm`, where `encrypter`, `decrypter` and `key_len` are the same code (based on a `ring::aead::Algorithm`) but `extract_keys` is different (because it produces different `ConnectionTrafficSecrets` variants). have replaced these free functions to different structs that each impl a new `KeyExtractor` trait that just does the `extract_keys()` part. that doesn't really make a lot of sense, though, because it implies any implementer of that trait is interchangeable with any other.

i think on balance i prefer how it was on the first round of this PR, where there were three structs that `impl Tls13AeadAlgorithm` and the shared parts were extracted out into a struct that owned the `ring::aead::Algorithm`.

Reply:
Okay, feel free to merge it like this or to revert it to your preferred approach.